prov/rxm/src/rxm_cq.c

```diff
@@ -916,23 +919,34 @@ static void rxm_cq_read_write_error(struct rxm_ep *rxm_ep)
 		return;
 	}
 
-	rx_buf = (struct rxm_rx_buf *)err_entry.op_context;
-
 	switch (RXM_GET_PROTO_STATE(err_entry.op_context)) {
 	case RXM_SAR_TX:
+		assert(err_entry.flags & FI_SEND);
+		sar_buf = err_entry.op_context;
+		err_entry.op_context = sar_buf->app_context;
+		err_entry.flags = ofi_tx_cq_flags(sar_buf->pkt.hdr.op);
+		goto set_tx;
 	case RXM_TX:
-	case RXM_INJECT_TX:
+		assert(err_entry.flags & FI_SEND);
+		eager_buf = err_entry.op_context;
+		err_entry.op_context = eager_buf->app_context;
+		err_entry.flags = ofi_tx_cq_flags(eager_buf->pkt.hdr.op);
+		goto set_tx;
 	case RXM_RNDV_TX:
-		util_cq = rxm_ep->util_ep.tx_cq;
-		if (rxm_ep->util_ep.flags & OFI_CNTR_ENABLED)
-			util_cntr = rxm_ep->util_ep.tx_cntr;
-		break;
+		assert(err_entry.flags & FI_SEND);
+		rndv_buf = err_entry.op_context;
+		err_entry.op_context = rndv_buf->app_context;
+		err_entry.flags = ofi_tx_cq_flags(rndv_buf->pkt.hdr.op);
+		goto set_tx;
 	case RXM_RNDV_ACK_SENT:
+		assert(err_entry.flags & FI_SEND);
+		goto set_rx;
 	case RXM_RX:
+		assert(err_entry.flags & FI_RECV);
+		goto set_rx;
 	case RXM_RNDV_READ:
-		util_cq = rx_buf->ep->util_ep.rx_cq;
-		util_cntr = rx_buf->ep->util_ep.rx_cntr;
-		break;
+		assert(err_entry.flags & FI_READ);
+		goto set_rx;
```

Comment:
The use of goto's in this function are unsightly.  Specifically, the goto set_rx is only there so that we can have assertions.  Please merge those case statements together and drop the set_rx goto. E.g.
```
assert((RXM_GET_PROTO_STATE(err_entry.op_context) == RXM_RNDV_ACK_SENT &&
    err_entry.flags & FI_SEND) ||
    (RXM_GET_PROTO_STATE(err_entry.op_context) == RXM_RX &&
    err_entry.flags & FI_RECV) ...)
rx_buf = (struct rxm_rx_buf *) err_entry.op_context;
util_cq = rx_buf->ep->util_ep.rx_cq;
util_cntr = rx_buf->ep->util_ep.rx_cntr;
err_entry.op_context = rx_buf->recv_entry->context;
err_entry.flags = rx_buf->recv_entry->comp_flags;
```

Reply:
👍 Done.